File: extension/src/tools/assignments.ts
/**
 * Assignments and the unified deadline view.
 * Ported from src/avenue_mcp/tools/assignments.py.
 *
 * `dropbox/folders/` may be instructor-only, so this degrades: if the folder
 * listing is denied, deadlines still come from the calendar, and the response
 * says plainly what was lost rather than silently returning less.
 *
 * THE RULE THAT COST THE MOST TO LEARN: submission status has THREE states,
 * not two. `mysubmissions` is 403 on the probed instances, and defaulting to
 * "not_submitted" on that silence tells a student they have not handed in work
 * they may well have. In the deadline view it was worse — items marked
 * submitted get filtered out by default, so a denied route silently emptied
 * "what's due this week". Both bugs shipped in the Python version and are
 * fixed there too.
 */

import { avenue } from "../avenue/client.js";
import { AvenueError } from "../avenue/errors.js";
import { addDays, daysUntil, describe, nowUtc, parseD2L, toUtcIso } from "../avenue/dates.js";
import { asFloat, asInt, isObj, pick, richText, toText, truncate } from "../avenue/models.js";
import { describeDenial } from "../institutions.js";
import { getCurrentInstitution } from "../settings.js";
import { calendarEvents } from "./calendar.js";
import { courseName } from "./context.js";
import { listCourses } from "./courses.js";

const INSTRUCTIONS_CAP = 2000;

/** Distinguishes "route denied" from "nothing submitted". */
const UNAVAILABLE = Symbol("submissions-unavailable");
type SubmissionResult = { submittedAt: Date | null; files: string[] } | null | typeof UNAVAILABLE;

export async function listAssignments(args: {
  org_unit_id: number;
  include_submitted?: boolean;
}) {
  const { org_unit_id } = args;
  const includeSubmitted = args.include_submitted ?? true;
  const inst = await getCurrentInstitution();

  let folders: unknown[];
  try {
    folders = await avenue.getPaged("le", `${org_unit_id}/dropbox/folders/`);
  } catch (err) {
    if (err instanceof AvenueError && err.kind === "PermissionDenied") {
      return degradedFromCalendar(org_unit_id, describeDenial(inst, "dropbox_folders"));
    }
    throw err;
  }

  const assignments = [];
  let statusUnavailable = false;

  for (const folder of folders) {
    if (!isObj(folder)) continue;
    const folderId = asInt(pick(folder, "Id", "FolderId"));
    if (folderId === null) continue;

    const due = parseD2L(pick(folder, "DueDate", "Due"));
    const instructions = toText(richText(pick(folder, "Instructions", "CustomInstructions")));

    const sub = await mySubmission(org_unit_id, folderId);
    let status: string;
    if (sub === UNAVAILABLE) {
      status = "unknown";
      statusUnavailable = true;
    } else if (sub !== null) {
      status = "submitted";
    } else {
      status = "not_submitted";
    }

    if (status === "submitted" && !includeSubmitted) continue;

    assignments.push({
      folder_id: folderId,
      name: String(pick(folder, "Name", "Title") ?? ""),
      due_date: describe(due),
      days_until_due: daysUntil(due),
      points_possible: asFloat(pick(folder, "OutOf", "PointsPossible", "TotalPoints")),
      instructions_text: truncate(instructions, INSTRUCTIONS_CAP).text || null,
      submission_status: status,
      submitted_at:
        sub !== UNAVAILABLE && sub !== null ? describe(sub.submittedAt) : null,
      submitted_files: sub !== UNAVAILABLE && sub !== null ? sub.files : [],
    });
  }

  assignments.sort(
    (a, b) =>
      (a.due_date?.utc ?? "9999").localeCompare(b.due_date?.utc ?? "9999") ||
      a.name.localeCompare(b.name),
  );

  return {
    org_unit_id,
    course_name: await courseName(org_unit_id),
    assignments,
    count: assignments.length,
    degraded: false,
    submission_status_available: !statusUnavailable,
    note: statusUnavailable
      ? `Submission status is not available — the learner submissions route was denied. ` +
        `${describeDenial(inst, "dropbox_mysubmissions")} Every assignment shows ` +
        `submission_status 'unknown': do NOT tell the user whether they have handed ` +
        `anything in. Due dates, points, and instructions above are accurate.`
      : null,
  };
}

async function degradedFromCalendar(orgUnitId: number, weight: string) {
  let events: Awaited<ReturnType<typeof calendarEvents>> = [];
  try {
    events = await calendarEvents(orgUnitId);
  } catch {
    events = [];
  }

  const items = events
    .filter((ev) => ev.kind === "assignment" || ev.kind === "event")
    .map((ev) => ({
      folder_id: null,
      name: ev.title,
      due_date: ev.due_date,
      days_until_due: daysUntil(ev.dueAt),
      points_possible: null,
      instructions_text: null,
      submission_status: "unknown",
      submitted_at: null,
      submitted_files: [],
    }));

  return {
    org_unit_id: orgUnitId,
    course_name: await courseName(orgUnitId),
    assignments: items,
    count: items.length,
    degraded: true,
    submission_status_available: false,
    note:
      `The assignment folder list could not be read, so point values, instructions, ` +
      `and submission status are unavailable. ${weight} These entries are ` +
      `calendar-derived — tell the user what is missing rather than inventing it.`,
  };
}

/**
 * The caller's own submissions.
 *
 * Returns UNAVAILABLE (not null) when the route is denied. Collapsing those
 * two is what produced the "you haven't submitted this" bug.
 */
async function mySubmission(orgUnitId: number, folderId: number): Promise<SubmissionResult> {
  let data: unknown;
  try {
    data = await avenue.get(
      "le",
      `${orgUnitId}/dropbox/folders/${folderId}/submissions/mysubmissions/`,
    );
  } catch (err) {
    if (err instanceof AvenueError) return UNAVAILABLE;
    throw err;
  }

  const entries = Array.isArray(data) ? data : isObj(data) ? [data] : [];
  let latest: { submittedAt: Date | null; files: string[] } | null = null;

  for (const entry of entries) {
    if (!isObj(entry)) continue;
    const subs = pick(entry, "Submissions");
    const records = Array.isArray(subs) ? subs : [entry];
    for (const rec of records) {
      if (!isObj(rec)) continue;
      const when = parseD2L(pick(rec, "SubmissionDate", "DateSubmitted", "Date"));
      if (latest === null || (when && (!latest.submittedAt || when > latest.submittedAt))) {
        const files = pick(rec, "Files");
        latest = {
          submittedAt: when,
          files: Array.isArray(files)
            ? files.filter(isObj).map((f) => String(pick(f, "FileName", "Name") ?? ""))
            : [],
        };
      }
    }
  }
  return latest;
}

// --- unified deadline view --------------------------------------------------

export async function getUpcomingDeadlines(args: {
  days_ahead?: number;
  include_submitted?: boolean;
}) {
  const daysAhead = args.days_ahead ?? 14;
  const includeSubmitted = args.include_submitted ?? false;
  const now = nowUtc();
  const windowEnd = addDays(now, daysAhead);

  const { courses } = await listCourses({});
  const deadlines: Array<Record<string, unknown>> = [];
  let anyStatus = false;

  for (const course of courses) {
    const oid = course.org_unit_id;

    // Calendar first: it covers quizzes and anything else an instructor dated.
    try {
      for (const ev of await calendarEvents(oid, { daysBack: 0, daysAhead })) {
        if (ev.dueAt < now || ev.dueAt > windowEnd) continue;
        deadlines.push({
          course_name: course.name,
          org_unit_id: oid,
          title: ev.title,
          type: ev.kind,
          due_date: ev.due_date,
          days_until: daysUntil(ev.dueAt, now),
          submission_status: "unknown",
        });
      }
    } catch {
      // One course's calendar failing must not cost the rest.
    }

    // Then assignment folders, which carry points and real status.
    try {
      for (const folder of await avenue.getPaged("le", `${oid}/dropbox/folders/`)) {
        if (!isObj(folder)) continue;
        const fid = asInt(pick(folder, "Id", "FolderId"));
        const due = parseD2L(pick(folder, "DueDate", "Due"));
        if (fid === null || !due || due < now || due > windowEnd) continue;

        const sub = await mySubmission(oid, fid);
        // THREE outcomes. UNAVAILABLE must not read as "submitted", or these
        // get filtered out below and the answer silently empties.
        let status: string;
        if (sub === UNAVAILABLE) status = "unknown";
        else if (sub !== null) status = "submitted";
        else status = "not_submitted";
        if (status !== "unknown") anyStatus = true;

        const title = String(pick(folder, "Name", "Title") ?? "");
        const existing = deadlines.find(
          (d) =>
            d["org_unit_id"] === oid &&
            String(d["title"] ?? "").trim().toLowerCase() === title.trim().toLowerCase(),
        );

        if (existing) {
          // Match on TITLE, never the timestamp alone: a quiz and an assignment
          // routinely share an 11:59 PM deadline, and matching on time made the
          // quiz inherit the assignment's "submitted" and vanish from the view.
          existing["submission_status"] = status;
          existing["points_possible"] = asFloat(pick(folder, "OutOf", "PointsPossible"));
          existing["folder_id"] = fid;
        } else {
          deadlines.push({
            course_name: course.name,
            org_unit_id: oid,
            title,
            type: "assignment",
            due_date: describe(due),
            days_until: daysUntil(due, now),
            submission_status: status,
            points_possible: asFloat(pick(folder, "OutOf", "PointsPossible")),
            folder_id: fid,
          });
        }
      }
    } catch {
      // Folder listing denied for this course; calendar entries still stand.
    }
  }

  const visible = includeSubmitted
    ? deadlines
    : deadlines.filter((d) => d["submission_status"] !== "submitted");

  visible.sort((a, b) => {
    const au = (a["due_date"] as { utc?: string } | null)?.utc ?? "9999";
    const bu = (b["due_date"] as { utc?: string } | null)?.utc ?? "9999";
    return au.localeCompare(bu) || String(a["course_name"]).localeCompare(String(b["course_name"]));
  });

  return {
    window_days: daysAhead,
    generated_at: toUtcIso(now),
    courses_checked: courses.length,
    deadlines: visible,
    count: visible.length,
    submission_status_available: anyStatus,
    note: anyStatus
      ? "Reflects the course calendar and assignment folders. Items an instructor never dated will not appear."
      : "Submission status is unavailable here, so work you have already handed in may still be listed. Report these as 'due', not as 'not done'.",
  };
}

File: src/avenue_mcp/tools/assignments.py
"""list_assignments and get_upcoming_deadlines (docs/03-mcp-tools.md).

get_upcoming_deadlines uses the CROSS-COURSE calendar route
(`GET /le/{v}/calendar/events/myEvents/?orgUnitIdsCSV=...`), which answers for
every course in one request. An earlier design fanned out one call per course
and was described as "the most request-heavy read tool"; it isn't one.

Both calendar variants require startDateTime and endDateTime. A bare call
returns 400, so `days_ahead` is converted into an explicit window rather than
passed through.
"""

from __future__ import annotations

import logging
from datetime import timedelta
from typing import Any

from avenue_mcp.client.d2l import D2LClient
from avenue_mcp.errors import PermissionDeniedError
from avenue_mcp.tools.courses import list_courses
from avenue_mcp.util.dates import (
    days_until,
    describe,
    parse_d2l_date,
    to_utc_param,
    utcnow,
)
from avenue_mcp.util.html import html_to_text

log = logging.getLogger(__name__)


async def _my_submissions(
    client: D2LClient, org_unit_id: int, folder_id: int
) -> list[dict[str, Any]]:
    """The learner-scoped submissions route. Degrades quietly to empty."""
    try:
        raw = await client.get_json(
            "le", f"{org_unit_id}/dropbox/folders/{folder_id}/submissions/mysubmissions/"
        )
    except Exception as exc:  # noqa: BLE001 - one folder must not kill the list
        log.debug("mysubmissions failed for folder %s: %s", folder_id, exc)
        return []
    return [s for s in (raw if isinstance(raw, list) else []) if isinstance(s, dict)]


async def list_assignments(
    client: D2LClient,
    *,
    org_unit_id: int,
    include_submitted: bool = True,
) -> dict[str, Any]:
    """A course's assignments with due dates, points, and submission status."""
    tz = client.settings.timezone

    try:
        folders = await client.get_json("le", f"{org_unit_id}/dropbox/folders/")
    except PermissionDeniedError:
        # The honest degraded shape: say what happened and point at the tool
        # that still works, rather than returning an empty list that reads
        # as "this course has no assignments".
        return {
            "org_unit_id": org_unit_id,
            "assignments": [],
            "count": 0,
            "assignments_available": False,
            "note": (
                "Assignment folders are not accessible from your account on this "
                "Avenue instance. Due dates may still be available via "
                "get_upcoming_deadlines, which reads the course calendar."
            ),
        }

    assignments: list[dict[str, Any]] = []
    for folder in folders if isinstance(folders, list) else []:
        if not isinstance(folder, dict):
            continue

        folder_id = folder.get("Id")
        due = parse_d2l_date(folder.get("DueDate"))

        submissions = (
            await _my_submissions(client, org_unit_id, int(folder_id))
            if folder_id is not None
            else []
        )
        submitted_at = None
        for submission in submissions:
            stamp = parse_d2l_date(submission.get("SubmissionDate"))
            if stamp and (submitted_at is None or stamp > submitted_at):
                submitted_at = stamp

        status = "submitted" if submissions else "not_submitted"
        if not include_submitted and status == "submitted":
            continue

        instructions = (folder.get("Instructions") or {}).get("Html") or ""

        assignments.append(
            {
                "folder_id": folder_id,
                "name": folder.get("Name") or "",
                "due_date": describe(due, tz),
                "days_until_due": days_until(due),
                "points_possible": folder.get("Assessment", {}).get("ScoreDenominator")
                if isinstance(folder.get("Assessment"), dict)
                else None,
                "instructions_text": html_to_text(instructions) or None,
                "submission_status": status,
                "submitted_at": describe(submitted_at, tz),
                "submission_count": len(submissions),
            }
        )

    assignments.sort(key=lambda a: (a["due_date"] or {}).get("utc") or "9999")
    return {
        "org_unit_id": org_unit_id,
        "assignments": assignments,
        "count": len(assignments),
        "assignments_available": True,
    }


def _event_type(event: dict[str, Any]) -> str:
    """Best-effort classification of a calendar event."""
    raw = " ".join(
        str(event.get(k) or "")
        for k in ("AssociationType", "EventType", "Title", "Description")
    ).lower()
    if "dropbox" in raw or "assignment" in raw:
        return "assignment"
    if "quiz" in raw or "test" in raw or "exam" in raw or "midterm" in raw:
        return "quiz"
    if "discussion" in raw:
        return "discussion"
    return "event"


async def get_upcoming_deadlines(
    client: D2LClient,
    *,
    days_ahead: int = 14,
    include_submitted: bool = False,
) -> dict[str, Any]:
    """Dated items across all active courses, soonest first."""
    tz = client.settings.timezone
    now = utcnow()
    window_end = now + timedelta(days=days_ahead)

    courses = (await list_courses(client))["courses"]
    if not courses:
        return {
            "window_days": days_ahead,
            "generated_at": describe(now, tz),
            "deadlines": [],
            "count": 0,
            "note": "No active courses found.",
        }

    names = {int(c["org_unit_id"]): c["name"] for c in courses}
    params = {
        "orgUnitIdsCSV": ",".join(str(i) for i in names),
        "startDateTime": to_utc_param(now),
        "endDateTime": to_utc_param(window_end),
    }

    # One request for every course. Falls back to per-course calls only if the
    # cross-course variant isn't available on this instance.
    try:
        events = await client.get_json("le", "calendar/events/myEvents/", params=params)
        source = "cross-course"
    except Exception as exc:  # noqa: BLE001
        log.info("Cross-course calendar unavailable (%s); falling back per course.", exc)
        events = []
        source = "per-course fallback"
        per_course = {
            "startDateTime": params["startDateTime"],
            "endDateTime": params["endDateTime"],
        }
        for org_unit_id in names:
            try:
                got = await client.get_json(
                    "le", f"{org_unit_id}/calendar/events/myEvents/", params=per_course
                )
            except Exception as inner:  # noqa: BLE001
                log.debug("Calendar failed for %s: %s", org_unit_id, inner)
                continue
            for event in got if isinstance(got, list) else []:
                if isinstance(event, dict):
                    event.setdefault("OrgUnitId", org_unit_id)
                    events.append(event)

    deadlines: list[dict[str, Any]] = []
    for event in events if isinstance(events, list) else []:
        if not isinstance(event, dict):
            continue
        due = parse_d2l_date(event.get("EndDateTime") or event.get("StartDateTime"))
        if due is None or due < now or due > window_end:
            continue

        org_unit_id = event.get("OrgUnitId") or event.get("OrgUnitIdentifier")
        try:
            org_unit_id = int(org_unit_id) if org_unit_id is not None else None
        except (TypeError, ValueError):
            org_unit_id = None

        deadlines.append(
            {
                "title": event.get("Title") or "",
                "course_name": names.get(org_unit_id, "") if org_unit_id else "",
                "org_unit_id": org_unit_id,
                "due_date": describe(due, tz),
                "days_until": days_until(due, now=now),
                "type": _event_type(event),
                "description": html_to_text(
                    (event.get("Description") or {}).get("Html")
                    if isinstance(event.get("Description"), dict)
                    else event.get("Description")
                )
                or None,
            }
        )

    deadlines.sort(key=lambda d: (d["due_date"] or {}).get("utc") or "")

    return {
        "window_days": days_ahead,
        "generated_at": describe(now, tz),
        "courses_covered": len(names),
        "source": source,
        "deadlines": deadlines,
        "count": len(deadlines),
        "note": (
            "Reflects the course calendar. Items your instructor didn't put on the "
            "calendar won't appear here."
        ),
    }
